### DISCUSSION TITLE
usb device class

### DISCUSSION THREAD

**Discussion by KozhinovAlexander** [CANnectivity/cannectivity#41] [Q&A] [answered]
I am trying to understand, why there are device/class and device_next/class paths with almost same code? In my understanding copy-pasting the code inceases testing effort exponentially.
UPD: Depending on longetivity of device/class support it may be easier to put same code pieces in one  .c file.

![grafik](https://github.com/user-attachments/assets/c7fd768e-6cdc-48fd-8170-05581217d66a)

**Comment by henrikbrixandersen** [comment, marked as answer]
> I am trying to understand, why there are device/class and device_next/class paths with almost same code?

As mentioned in https://github.com/CANnectivity/cannectivity/blob/main/README.md, CANnectivity supports both the classic and the experimental USB `device_next` stacks provided by Zephyr. While the files may seem very similar on quick glance, the similarities end when it comes to the USB device APIs, they need to implement.

> In my understanding copy-pasting the code inceases testing effort exponentially.

If you had taken a look in e.g. https://github.com/CANnectivity/cannectivity/tree/main/tests/subsys/usb/gs_usb/host, you would have seen that the testing efforts are not duplicated between the two class implementations. The only difference in the test code relates to the initialisation of the USB stack. Both stacks can be tested on the same hardware, both for test cases and firmware application.
